ruskeyverbs/views.py

```diff
@@ -26,19 +51,618 @@ def index(request):
 
 @login_required
 def VerbListPerUser(request):
-    # get list of all available verbs sorted by due date for given user
-    verbs_with_due_dates = Verb.objects.annotate(due=Min('example__performanceperexample__due_date', filter=Q(example__performanceperexample__user_id=request.user.pk))).order_by('due').exclude(due=None)
-    verbs_without_due_dates = Verb.objects.annotate(due=Min('example__performanceperexample__due_date', filter=Q(example__performanceperexample__user_id=request.user.pk))).exclude(~Q(due=None))
-    sorted_verb_list = list(chain(verbs_with_due_dates, verbs_without_due_dates))
-    # sorted_verb_list = sorted(Verb.objects.all(),
-    #                           key=lambda a: a.get_earliest_due_date(request.user))
-    current_user = request.user
+    """
+    list view for verbs available in the system. Sorts by due date
+    (if studied previously) then frequency ranking
+    """
+    # two query sets because otherwise unstudied verbs would display before
+    # studied ones
+    verbs_with_due_dates = Verb.objects.annotate(
+        due=Min(
+            'example__performanceperexample__due_date',
+            filter=Q(
+                example__performanceperexample__user_id=request.user.pk)
+            )
+        ).order_by('due').exclude(due=None)
+    verbs_without_due_dates = Verb.objects.annotate(
+        due=Min('example__performanceperexample__due_date',
+                filter=Q(
+                    example__performanceperexample__user_id=request.user.pk)
+                )
+        ).exclude(~Q(due=None))
+    sorted_verb_list = list(chain(verbs_with_due_dates,
+                                  verbs_without_due_dates))
     return render(
         request,
         'ruskeyverbs/list_view_per_user.html',
-        context={'verb_list': sorted_verb_list, 'current_user': current_user}
+        context={'verb_list': sorted_verb_list}
     )
 
 
-class VerbDetails(LoginRequiredMixin, generic.DetailView):
-    model = Verb
+@login_required
+def VerbDetails(request, pk):
+    """
+    detail view for verbs; page will display conjugation and example info/audio
+    for user
+    """
+    verb = Verb.objects.get(pk=pk)
+    example_pk = Example.objects.annotate(
+        due=Min('performanceperexample__due_date',
+                filter=Q(
+                    performanceperexample__user_id=request.user.pk)
+                )
+        ).filter(verb=verb).order_by('due')[0].pk
+    """
+    If the user has previously studied the example, send them right
+    into the quiz; otherwise, send them to the study page before the
+    quiz
+    """
+    try:
+        print(PerformancePerExample.objects.get(pk=example_pk,
```

Comment:
The print statement is not necessary here. And it is probably a little cleaner to do an exists() query against the database rather try to fetch and catch the error.

Reply:
Swapped exists() calls for all of the try: except: blocks testing for objects.

Reply:
Good